### PR TITLE
PostgreSQL 12 and PostGIS 3 compatibility

### PR DESCRIPTION
**Pull request by manmorjim**
Changes in cartodb to be compatible with PostgreSQL 12 and PostGIS 3.

### CODE REVIEW

**Comment by skgsergio** [comment]
Missing extension version update in `carto-package.json`

File: services/user-mover/export_user.rb [hunk 1]
```diff
@@ -185,7 +194,13 @@ def get_org_metadata(organization_id)
       end
 
       def get_org_users(organization_id)
-        q = pg_conn.exec("SELECT * FROM users WHERE organization_id = '#{organization_id}'")
+        # owners have to be in the last position to prevent errors
```
Comment (alrocar) [inline services/user-mover/export_user.rb]: 👍 to meaningful comments

File: services/user-mover/legacy_functions.rb [hunk 2]
```diff
@@ -27,15 +27,24 @@ module LegacyFunctions
...
-        'AGGREGATE st_union(raster)',
-        'AGGREGATE st_union(raster,integer)',
-        'AGGREGATE st_union(raster,integer,text)',
-        'AGGREGATE st_union(raster,text)',
-        'AGGREGATE st_union(raster,text,text)',
-        'AGGREGATE st_union(raster,text,text,text)',
-        'AGGREGATE st_union(raster,text,text,text,double precision)',
-        'AGGREGATE st_union(raster,text,text,text,double precision,text,text,text,double precision)',
-        'AGGREGATE st_union(raster,text,text,text,double precision,text,text,text,double precision,text,text,text,double precision)',
+         # PG12_DEPRECATED
```
Comment (alrocar) [inline services/user-mover/legacy_functions.rb]: are these comments in every line for a reason?

File: spec/lib/carto/ghost_tables_manager_spec.rb [hunk 3]
```diff
@@ -229,33 +229,35 @@ def run_in_user_database(query)
...
     it 'should link raster tables' do
-      run_in_user_database(%{
-        CREATE TABLE manolo_raster ("cartodb_id" uuid, "the_raster_webmercator" raster);
-        CREATE TRIGGER test_quota_per_row
-          BEFORE INSERT OR UPDATE
-          ON manolo_raster
-          FOR EACH ROW
-          EXECUTE PROCEDURE cdb_checkquota('0.001', '-1', 'public');
-      })
+      if ::User[@user.id].in_database.table_exists?('raster_overviews')
```
Comment (alrocar) [inline spec/lib/carto/ghost_tables_manager_spec.rb]: return unless... ☝️

File: spec/models/carto/oauth_app_user_spec.rb [hunk 4]
```diff
@@ -185,22 +185,22 @@ module Carto
         oau = OauthAppUser.create!(user: @carto_user, oauth_app: @app, scopes: [o1, o2])
 
         oau.upgrade!([])
-        expect(oau.scopes.sort).to(eq([o1, o2].sort))
+        expect(oau.scopes).to(match_array([o1, o2]))
```
Comment (alrocar) [inline spec/models/carto/oauth_app_user_spec.rb]: ❤️ 

I added the sort recently, but much better the `match_array`

File: spec/models/carto/user_migration_api_key_spec.rb [hunk 5]
```diff
@@ -18,30 +18,32 @@
...
-    user.in_database.execute('INSERT INTO i_hate_raster VALUES(ST_MakeEmptyRaster(100, 100, 0, 0, 100, 100, 0, 0, 2274))')
-    user.in_database.execute("UPDATE i_hate_raster SET rast = ST_AddBand(rast, 1, '32BF'::text, 0)")
-    user.in_database.execute("UPDATE i_hate_raster SET rast = ST_AddBand(rast, 1, '32BF'::text, 0)")
-    user.in_database.execute("SELECT AddRasterConstraints('i_hate_raster', 'rast')")
-    user.in_database.execute("SELECT ST_CreateOverview('i_hate_raster'::regclass, 'rast', 2)")
-    user.in_database.execute('DROP TABLE i_hate_raster')
-    export = Carto::UserMigrationExport.create(user: carto_user, export_metadata: true)
-    export.run_export
-    user.destroy
+    if user.in_database.table_exists?('raster_overviews')
```
Comment (alrocar) [inline spec/models/carto/user_migration_api_key_spec.rb]: ☝️

File: spec/models/table_spec.rb [hunk 6]
```diff
@@ -1362,16 +1362,18 @@ def expect_save_to_fail_validation(table)
       end
 
       it "should not fail when the analyze is executed in update_table_geom_pg_stats and raises a PG::UndefinedColumn" do
-        delete_user_data @user
-        data_import = DataImport.create(user_id: @user.id,
-                                        data_source: fake_data_path('import_raster.tif.zip'))
-        data_import.run_import!
+        if @user.in_database.table_exists?('raster_overviews')
```
Comment (alrocar) [inline spec/models/table_spec.rb]: ☝️

File: app/models/user/db_service.rb [hunk 7]
```diff
@@ -1522,18 +1526,34 @@ def create_function_invalidate_varnish_http
...
                 retry = #{varnish_retry}
                 trigger_verbose = #{varnish_trigger_verbose}
-                for i in ('httplib', 'base64', 'hashlib'):
+
+                if not 'httplib' in GD:
+                  try:
+                    GD['httplib'] = __import__('httplib')
+                  except:
+                    from http import client
+                    GD['httplib'] = client
```
[the remark this answers is not in the document]
Comment (manmorjim) [reply app/models/user/db_service.rb]: Yes, that's right. AFAIK we're not going to end support for PG11 (which uses python2) yet. 

However, I'll add a `# PG12_DEPRECATED` comment in order to get a quick access to that code we have to adapt when we remove the PG11 support.

File: services/importer/spec/acceptance/raster2pgsql_spec.rb [hunk 8]
```diff
@@ -128,13 +131,16 @@
...
+    # PG12_DEPRECATED
+    if @user.in_database.table_exists?('raster_columns')
+      metadata = @user.in_database.fetch(%{
+          SELECT * FROM raster_columns
+          WHERE r_table_schema = '#{job.schema}' AND r_table_name = '#{job.table_name}'
+        }).first
+      metadata[:srid].should eq 4326
+      metadata[:scale_x].should be_within(TOLERANCE).of +0.148148148148133
+      metadata[:scale_y].should be_within(TOLERANCE).of -0.148148148148133
+    end
```
[the remark this answers is not in the document]
Comment (manmorjim) [reply services/importer/spec/acceptance/raster2pgsql_spec.rb]: :fire: :fire: :fire: :fire:

File: services/user-mover/legacy_functions.rb (hunk 2 above)
Comment (manmorjim) [reply services/user-mover/legacy_functions.rb]: those lines are using raster which is deprecated in PG12. I tried to remove them but a lot of errors came up so it is not as easy as I thought at the beginning. 

By that reason I've just marked as deprecated, to make easier for us to find and remove them in a future.

But if you think this is just adding noise or does not help us I can revert this file :)

Comment (alrocar) [reply services/user-mover/legacy_functions.rb]: That's fine, I was wondering what were they for 😅

**Comment by alrocar** [comment]
@manmorjim I see CI is failing due to the cartodb extension, not sure if there's a change to make it ✅

**Comment by manmorjim** [comment]
> @manmorjim I see CI is failing due to the cartodb extension, not sure if there's a change to make it

Yes, I'm talking with infra about this because I think we have to create the docker image of `Dockerfile.cartodb.postgres12.base` again.

**Comment by Algunenano** [comment]
> Yes, I'm talking with infra about this because I think we have to create the docker image of Dockerfile.cartodb.postgres12.base again.

For CI, It should be enough with updating the submodule. That's how it used to be in the old CI, so if it isn't now then something needs to be done. We can't expect to generate a new docker image every time we want to test a change.

**Pull request merged**